src/conversions/num_bigint.rs

```diff
@@ -138,53 +145,66 @@ macro_rules! bigint_conversion {
-
-                    let n_bytes = {
+                    let n_bits = {
                         cfg_if::cfg_if! {
                             if #[cfg(not(Py_LIMITED_API))] {
                                 // fast path
                                 let n_bits = ffi::_PyLong_NumBits(num.as_ptr());
                                 if n_bits == (-1isize as usize) {
                                     return Err(crate::PyErr::fetch(py));
-                                } else if n_bits == 0 {
-                                    0
-                                } else {
-                                    (n_bits - 1 + $is_signed) / 8 + 1
                                 }
+                                n_bits
                             } else {
                                 // slow path
                                 let n_bits_obj = num.getattr(py, crate::intern!(py, "bit_length"))?.call0(py)?;
                                 let n_bits_int: &PyLong = n_bits_obj.downcast_unchecked(py);
-                                let n_bits = n_bits_int.extract::<usize>()?;
-                                if n_bits == 0 {
-                                    0
-                                } else {
-                                    (n_bits - 1 + $is_signed) / 8 + 1
-                                }
+                                n_bits_int.extract::<usize>()?
                             }
                         }
                     };
 
-                    if n_bytes <= 128 {
-                        let mut buffer = [0; 128];
-                        extract(num.as_ref(py), &mut buffer[..n_bytes], $is_signed)?;
-                        Ok($from_bytes(&buffer[..n_bytes]))
-                    } else {
-                        let mut buffer = vec![0; n_bytes];
-                        extract(num.as_ref(py), &mut buffer, $is_signed)?;
-                        Ok($from_bytes(&buffer))
+                    if n_bits == 0 {
+                        return Ok(0.try_into().unwrap())
                     }
+
+                    let n_digits = (n_bits + 31 + $is_signed) / 32;
+
+                    let mut buffer = extract(
+                        num.as_ref(py),
+                        n_digits,
+                        $is_signed
+                    )?;
+
+                    buffer
+                        .iter_mut()
+                        .for_each(|chunk|{*chunk = u32::from_le(*chunk)});
+
+                    Ok($from_digits(buffer))
                 }
             }
         }
     };
 }
-bigint_conversion!(BigUint, 0, BigUint::to_bytes_le, BigUint::from_bytes_le);
+bigint_conversion!(
+    BigUint,
+    0,
+    BigUint::to_bytes_le,
+    //from digits conversion
+    BigUint::new
+);
 bigint_conversion!(
     BigInt,
     1,
     BigInt::to_signed_bytes_le,
-    BigInt::from_signed_bytes_le
+    //from digits conversion
+    |mut digits: Vec<u32>| {
+        if digits.last().unwrap() >> 31 != 0 {
```

Comment:
I wonder if this loop could be merged with the `u32::from_le` loop based on the value of `$is_signed` to avoid stashing code down here?

Reply:
I don't quite know how, since this loop is for negative to BigUint conversion

Reply:
But it is applied only if `$is_signed` isn't? So `$from_digits` could always be `BigUint::new`/`BigInt::new` and the conversion could happen inside the code using e.g.

```rust
if $is_signed && digits.last().unwrap() >> 31 != 0 {
  buffer.iter().for_each(|chunk| *chunk = !u32::from_le(*chunk));
} else {
  buffer.iter().for_each(|chunk| *chunk = u32::from_le(*chunk));
}
```

giving us a single loop and the option to pipeline the negation and byte swap.

Reply:
I guess it could be
```
if digits.last().unwrap() >> 31 != 0 {
  if $is_signed {
    buffer.iter().for_each(|chunk| *chunk = !u32::from_le(*chunk));
  } else {
    buffer.iter().for_each(|chunk| *chunk = u32::from_le(*chunk));
  }
}
```
But then how would I add/subtract one to the BigInt to finish the conversion from two's complement?

Reply:
> But then how would I add/subtract one to the BigInt to finish the conversion from two's complement?

This is complicated to do on the digits because it might carry from the least up to the most significant digit, right?

Reply:
Yup, exactly!

Reply:
I know it is a bad habit when maintainers ask for unrelated clean-ups but since this is an optimization to begin, maybe can we avoid the macro entirely and use the same approach `num-bigint` internally, i.e. we directly implement `impl FromPyObject for BigUint` and then implement `impl FromPyObject for BigInt` in terms of that?

That would still leave us with two loops, but hopefully cleaner code that does not have to work in "pseudo-generic context."

Reply:
Are you okay with some noticeable code duplication then? I would love to get rid of the macros and have a language server help me write code (it's pretty limited for macros) but that would lead to copypasting some functions for BigInt and BigUInt separately.

Reply:
oh wait you meant only `impl FromPyObject` and not the whole thing... oops!

Reply:
> oh wait you meant only impl FromPyObject and not the whole thing... oops!

That said, I would not mind a bit of code duplication to get completely rid of the macro either, as long as the meaty part which does the conversion to a Python object is factored out into a helper function similar to the existing `extract`. I think writing a bit of mindless boiler plate is a reasonable price to pay for avoiding the macro.

Reply:
> I would not mind a bit of code duplication to get completely rid of the macro either
i couldnt come up with a sensible way to refactor it, all options smelled worse than the macro (to me personally, im a data scientist who mostly writes python so take it with a grain of salt), so i decided to leave it to people that are actually familiar with the code style and conventions. I did refactor the `impl FromPyObject` and left it at that